domain/port/service/watcher.go

```diff
@@ -0,0 +1,166 @@
+		mapper eventsource.Mapper,
+	) (watcher.StringsWatcher, error)
+
+	// NewNamespaceNotifyMapperWatcher returns a new namespace notify watcher
+	// for events based on the input change mask and mapper.
+	NewNamespaceNotifyMapperWatcher(
+		namespace string, changeMask changestream.ChangeType, mapper eventsource.Mapper,
+	) (watcher.NotifyWatcher, error)
+}
+
+// WatcherState describes the methods that the service needs for it's watchers.
+type WatcherState interface {
+	// WatchOpenedPortsTable returns the name of the table that should be watched
+	WatchOpenedPortsTable() string
+
+	// InitialWatchOpenedPortsStatement returns the query to load the initial event
+	// for the WatchOpenedPorts watcher
+	InitialWatchOpenedPortsStatement() string
+
+	// GetMachinesForEndpoints returns map from endpoint uuids to the uuids of
+	// the machines which host that endpoint for each provided endpoint uuid.
+	GetMachinesForEndpoints(ctx context.Context, endpointUUIDs []string) (map[string]string, error)
+
+	// GetApplicationsForEndpoints returns map from endpoint uuids to the uuids
+	// of the applications which host that endpoint for each provided endpoint uuid.
+	GetApplicationsForEndpoints(ctx context.Context, endpointUUIDs []string) (map[string]string, error)
+}
+
+// WatchOpenedPorts returns a strings watcher for opened ports. This watcher
+// will emit events for changes to the opened ports table. The events will
+// contain the machine UUIDs that are associated with the port range.
+func (s *WatchableService) WatchOpenedPorts(ctx context.Context) (watcher.StringsWatcher, error) {
+	return s.watcherFactory.NewNamespaceMapperWatcher(
+		s.st.WatchOpenedPortsTable(),
+		changestream.Create|changestream.Delete,
+		eventsource.InitialNamespaceChanges(s.st.InitialWatchOpenedPortsStatement()),
+		s.portRangeToMachineMapper,
+	)
+}
+
+// WatchOpenedPortsForApplication returns a notify watcher for opened ports. This
+// watcher will emit events for changes to the opened ports table that are associated
+// with the given application
+func (s *WatchableService) WatchOpenedPortsForApplication(ctx context.Context, applicationUUID string) (watcher.NotifyWatcher, error) {
+	return s.watcherFactory.NewNamespaceNotifyMapperWatcher(
+		s.st.WatchOpenedPortsTable(),
+		changestream.Create|changestream.Delete,
+		s.filterForApplication(applicationUUID),
+	)
+}
+
+// portRangeToMachineMapper is a mapper that maps port range events to machine UUIDs
+// of the machine hosting that endpoint
+func (s *WatchableService) portRangeToMachineMapper(
+	ctx context.Context, db database.TxnRunner, events []changestream.ChangeEvent,
+) ([]changestream.ChangeEvent, error) {
+
+	endpointUUIDs := transform.Slice(events, func(e changestream.ChangeEvent) string {
+		return e.Changed()
+	})
+
+	endpointToMachines, err := s.st.GetMachinesForEndpoints(ctx, endpointUUIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	newEvents := make([]changestream.ChangeEvent, len(events))
```

Comment:
You don't want to emit duplicate machine UUIDs here, so this may not be an equal length slice.

If multiple endpoints/units are changing on a machine, you also don't know which `changeType` applies to the machine. So change this in order that:
- `GetMachinesForEndpoints` does not return the endpoint UUID. It should be a `SELECT DISTINCT`.
- Don't iterate and do a map lookup, just create a new slice of shims for those machines with _update_ as the change.

Reply:
Yes good point. I was working under the assumption that the events had to maintain their order, an order that would be lost in a SQL query. However, now i think about it this obviously doesn't matter